[fix bug 1267658] Create shared FxA iframe JS lib.

jpetto opened the pull request:
## Description

DRY up our JS code that facilitates embedding the FxA iframe on bedrock pages.

## Bugzilla link

[bug 1267658](https://bugzilla.mozilla.org/show_bug.cgi?id=1267658)

## Testing

~~**Despite 2 successful pushes, demo3 is throwing 500 errors. Will investigate further tomorrow.** (Demo links below are broken until resolved.)~~

**Demo 3 links below are now functional.**

Using a [profile configured for FxA on our demo servers](http://bedrock.readthedocs.io/en/latest/firefox-accounts.html#demo-server-testing), test the following demo3 URLs:

- [`/firefox/47.0/firstrun/`](https://www-demo3.allizom.org/en-US/firefox/47.0/firstrun/)
- [`/firefox/47.0/firstrun/?v=6`](https://www-demo3.allizom.org/en-US/firefox/47.0/firstrun/?v=6)
- [`/firefox/accounts/`](https://www-demo3.allizom.org/en-US/firefox/accounts/)

## Checklist
- [ ] Related functional & integration tests passing.

jpetto commented:
Marked as **Do Not Merge** until we get sign-off from garethc on GA changes.

alexgibson commented on docs/firefox-accounts.rst:
```diff
@@ -84,4 +84,28 @@ Clearing the iframe cache
+
+#. Add the FxA JavaScript & Less files to the page's bundles:
+
+    - ``media/js/base/mozilla-fxa-iframe.js``
```
Nit: no need to indent this list?

jpetto replied:
Hm, I thought it needed to be indented to be a sub-list. rst drives me nuts. 😀

alexgibson commented on docs/firefox-accounts.rst:
```diff
@@ -84,4 +84,28 @@ Clearing the iframe cache
+
+    ``<iframe id="fxa" scrolling="no" data-src="{{ settings.FXA_IFRAME_SRC }}?utm_campaign=fxa-embedded-form&amp;utm_content=fx-{{ version }}&amp;service=sync&amp;context=iframe&amp;style=chromeless&amp;haltAfterSignIn=true"></iframe>``
+
+    Note that each implementation of the ``<iframe>`` may require unique URL
```
Could use a `::note` block here instead?

jpetto replied:
Ah, good call.

alexgibson commented on tests/unit/spec/base/mozilla-fxa-iframe.js:
```diff
@@ -0,0 +1,187 @@
+            Mozilla.Client.FirefoxMajorVersion = fxVersion;
+        });
+
+        it('should append URL encoded email if sent in config', function () {
```
Would it be worth adding another test here to make sure invalid email addresses aren't encoded in the URL?

jpetto replied:
Sounds like a good test to me. Will do. 👍

alexgibson commented on media/js/base/mozilla-fxa-iframe.js:
```diff
@@ -0,0 +1,205 @@
+        // remember iframe has loaded
+        _handshake = true;
+
+        if (typeof _config.onPing === 'function') {
```
Would it be worth adding a small utility `_isFunction()` method to avoid repeating the condition here and below? Something like

    function _isFunction(item) {
        return typeof item === 'function';
    }

jpetto replied:
Good idea. Will implement.

alexgibson commented:
Looks like `firstrun.js` and `fxa-complex.js` are now identical. Could we cull one and use a common file to init the iframe?

alexgibson commented:
Few small suggestions, but code-wise this looks great. Yay for tests!

I'll wait until our demo3 issues are resolved to give this a test functionally before giving the final r+.

jpetto commented:
@alexgibson - Addressed review comments, and finally got demo3 working (with jgmize's help). [Bug has a needinfo for garethc to review GA](https://bugzilla.mozilla.org/show_bug.cgi?id=1267658#c2) - hopefully will be verified soon.

alexgibson commented:
Tested all three URL's on demo3 via signing up for a new account and all worked as expected. Code changes look good too, so it's an r+ from me to squash & merge when ready 👍

jpetto commented:
Squashed!

jpetto commented:
pgerman has given us the GA green light, so this is good to merge. 🎆
